auth0/v3/authentication/token_verifier.py

```diff
@@ -0,0 +1,225 @@
+        except jwt.exceptions.InvalidSignatureError:
+            raise TokenValidationError("Invalid token signature.")
+        return decoded
+
+
+class SymmetricSignatureVerifier(SignatureVerifier):
+
+    def __init__(self, shared_secret, algorithm="HS256"):
+        SignatureVerifier.__init__(self, algorithm)
+        self._shared_secret = shared_secret
+
+    def _fetch_key(self, key_id=None):
+        return self._shared_secret
+
+
+class AsymmetricSignatureVerifier(SignatureVerifier):
+
+    def __init__(self, jwks_url, algorithm="RS256", _jwks_fetcher=None):
+        SignatureVerifier.__init__(self, algorithm)
+        self._fetcher = _jwks_fetcher or JwksFetcher(jwks_url)
+
+    def _fetch_key(self, key_id=None):
+        return self._fetcher.get_key(key_id)
+
+
+class JwksFetcher():
+    CACHE_TTL = 600  # 10 min cache lifetime
+
+    def __init__(self, jwks_url, cache_ttl=CACHE_TTL):
+        self._jwks_url = jwks_url
+        self._init_cache(cache_ttl)
+        return
+
+    def _init_cache(self, cache_ttl):
+        self._cache_value = {}
+        self._cache_date = 0
+        self._cache_ttl = cache_ttl
+        self._cache_is_fresh = False
+
+    def _fetch_jwks(self, force=False):
+        has_expired = self._cache_date + self._cache_ttl < time.time()
+
+        if not force and not has_expired:
+            # Return from cache
+            self._cache_is_fresh = False
+            return self._cache_value
+
+        # Invalidate cache and fetch fresh data
+        self._cache_value = {}
+        response = requests.get(self._jwks_url)
+
+        if response.ok:
+            # Update cache
+            jwks = response.json()
+            self._cache_value = self._parse_jwks(jwks)
+            self._cache_is_fresh = True
+            self._cache_date = time.time()
+        return self._cache_value
+
+    @staticmethod
+    def _parse_jwks(jwks):
+        keys = {}
+
+        for key in jwks['keys']:
+            # noinspection PyUnresolvedReferences
+            # requirement already includes cryptography -> pyjwt[crypto]
+            rsa_key = jwt.algorithms.RSAAlgorithm.from_jwk(json.dumps(key))
+            keys[key["kid"]] = rsa_key
+        return keys
+
+    def get_key(self, key_id):
+        keys = self._fetch_jwks()
+
+        if keys and key_id in keys:
+            return keys[key_id]
+
+        if not self._cache_is_fresh:
+            keys = self._fetch_jwks(force=True)
+            if keys and key_id in keys:
+                return keys[key_id]
+        raise TokenValidationError('RSA Public Key with ID "{}" was not found.'.format(key_id))
+
+
+class TokenVerifier():
+
+    def __init__(self, signature_verifier, issuer, audience, leeway=0, _clock=None):
+        if not signature_verifier or not isinstance(signature_verifier, SignatureVerifier):
+            raise TypeError("signature_verifier must be an instance of SignatureVerifier.")
+
+        self._options = {
```

Comment:
I'm not proud of this object here. What I tried to achieve is putting every parameter on an instance object so that I can access them later in the `verify` method. Any ideas are welcome :)

Reply:
```python
def __init__(self, **kwargs):
    self.__dict__.update(kwargs)
    # self.signature_verifier, self.issuer, etc...
```
If you want to make it less verbose